src/dawn/Optimizer/PassFieldVersioning.cpp

```diff
@@ -293,4 +328,86 @@ PassFieldVersioning::fixRaceCondition(const iir::DependencyGraphAccesses* graph,
   return RCKind::RK_Fixed;
 }
 
+bool PassFieldVersioning::checkReadBeforeWrite(int AccessID,
+                                               const std::unique_ptr<iir::MultiStage>& mss) {
+
+  for(const auto& stmtAccessesPair : iterateIIROver<iir::StatementAccessesPair>(*mss)) {
+
+    // Find first if this statement has a read
+    auto readAccessIterator =
+        stmtAccessesPair->getCallerAccesses()->getReadAccesses().find(AccessID);
```

Comment:
unfortunately this algorithm is not mathematically complete. The following example
```
vertical_region(k_start, k_start)
 a[k]=0;
vertical_region(k_start+1, k_end)
a[k] = a[k-1]; 
```
will trigger a false positive.

We worked on a complete algorithm for the caches. 
MultiStages::computeReadAccessInterval will return a MultiInterval where there will be an access to main memory because there is a read before write. So you just have to use and if the multiinterval is non empty, then there is a read before write. You could decide to trigger a copy just for the selected intervals  of the multiinterval...